ibis-server/app/util.py

```diff
@@ -101,7 +114,36 @@ def build_context(headers: Header) -> Context:
     return extract(headers)
 
 
+def append_fallback_context(headers: Header, span: trace.Span) -> Headers:
+    if headers is None:
+        headers = {}
+    else:
+        headers = dict(headers)
+    span = NonRecordingSpan(span.get_span_context())
+    context = set_span_in_context(span)
+    # https://opentelemetry.io/docs/languages/python/propagation/
+    W3CBaggagePropagator().inject(headers, context)
+    TraceContextTextMapPropagator().inject(headers, context)
+    return Headers(headers)
+
+
 @tracer.start_as_current_span("pushdown_limit", kind=trace.SpanKind.INTERNAL)
 def pushdown_limit(sql: str, limit: int | None) -> str:
     ctx = wren_core.SessionContext()
     return ctx.pushdown_limit(sql, limit)
+
+
+def get_fallback_message(
+    logger, prefix: str, datasource: DataSource, mdl_hash: str, sql: str
+) -> str:
+    if sql is not None:
+        sql = sql.replace("\n", " ")
+
+    message = orjson.dumps(
+        {"datasource": datasource, "mdl_hash": mdl_hash, "sql": sql}
```

Comment:
I think we encode the mdl in base64, so maybe we should just name it `mdl` or `mdl_base64`.